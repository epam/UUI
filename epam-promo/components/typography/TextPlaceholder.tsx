import * as React from 'react';
import css from './TextPlaceholder.scss';
import cx from 'classnames';
import { PropsWithChildren } from 'react';
import { IHasRawProps } from "@epam/uui-core";

export interface TextPlaceholderProps extends IHasRawProps<React.HTMLAttributes<HTMLDivElement>> {
    /** Number of fake 'words to show */
    wordsCount?: number;
    /** Placeholder's color */
    color?: 'gray10' | 'gray40';
    /** Disables animation */
    isNotAnimated?: boolean;
}

export const TextPlaceholder: React.FunctionComponent<PropsWithChildren<TextPlaceholderProps>> = (props) => {
    const pattern = `0`;
    const text = React.useMemo(() => {
        const words = [];
        for (let i = 0; i < (props.wordsCount || 1); i++) {
            let lengthWord = Math.floor(Math.random() * 10 + 8);
            words.push(pattern.repeat(lengthWord));
        }
        return words;
    }, [props.wordsCount]);


    function renderText() {
        return text.map((it: string, index: number) => (
            <span
                suppressHydrationWarning={ true }
                key={ index }
                className={ cx([
                    css.loadingWord,
                    css['text-placeholder-color-' + (props.color || 'gray40')],
                    !props.isNotAnimated && css.animatedLoading,
                ]) }
                dangerouslySetInnerHTML={ {__html: it} }
            />
        ));
    }

    return (
        <div aria-busy={ true } className={ css.container } { ...props.rawProps }>
            { renderText() }
        </div>
    );
};
