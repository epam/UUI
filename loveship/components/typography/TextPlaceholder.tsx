import React from 'react';
import cx from 'classnames';
import { IHasCX, IHasRawProps } from '@epam/uui-core';
import * as types from '../types';
import styles from '../../assets/styles/scss/loveship-color-vars.scss';
import css from './TextPlaceholder.scss';

export interface TextPlaceholderProps extends IHasCX, IHasRawProps<React.HTMLAttributes<HTMLDivElement>> {
    wordsCount?: number;
    color?: types.EpamColor;
    isNotAnimated?: boolean;
}

export const TextPlaceholder = React.forwardRef<HTMLDivElement, TextPlaceholderProps>((props, ref) => {
    const pattern = `&nbsp;`;
    const text = React.useMemo(() => {
        const words = [];
        for (let i = 0; i < (props.wordsCount || 1); i++) {
            const lengthWord = Math.floor(Math.random() * 10 + 8);
            words.push(pattern.repeat(lengthWord));
        }
        return words;
    }, [props.wordsCount]);

    return (
        <div ref={ ref } aria-busy={ true } className={ css.container } { ...props.rawProps }>
            { text.map((it, index) => (
                <span
                    suppressHydrationWarning={ true }
                    key={ index }
                    dangerouslySetInnerHTML={ { __html: it } }
                    className={ cx([
                        props.cx,
                        css.loadingWord,
                        styles[`color-${props.color || 'night100'}`],
                        !props.isNotAnimated && css.animatedLoading,
                    ]) }
                />
            )) }
        </div>
    );
});
