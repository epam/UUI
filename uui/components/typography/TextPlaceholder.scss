@keyframes skeleton_loading {
    0% { background-position: 200% 0; }
    100% { background-position: 0 0; }
}

.container {
    display: flex;
    flex-wrap: wrap;
}

.loading-word {
    user-select: none;
    font-family: var(--font-redacted);
    letter-spacing: -1px;
    margin: 0 3px 3px 0;
    color: var(--uui-text-placeholder);
    vertical-align: 1px; // to fix vertical over-sizing of line
    &:last-child {
        margin-right: 0;
    }

    &.animated-loading {
        color: transparent;
        background: var(--uui-text-placeholder-gradient);
        background-size: 200% 100%;
        background-clip: border-box;
        -webkit-background-clip: text;
        animation: skeleton_loading 1s ease infinite;
    }
}
