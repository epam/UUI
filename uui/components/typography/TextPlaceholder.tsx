import * as React from 'react';
import cx from 'classnames';
import css from './TextPlaceholder.scss';
import '../../assets/styles/variables/typography/textPlaceholder.scss';

export interface TextPlaceholderProps {
    wordsCount?: number;
    isNotAnimated?: boolean;
}

export const TextPlaceholder: React.FunctionComponent<TextPlaceholderProps> = props => {
    const pattern = `0`;
    const text = React.useMemo(() => {
        const words = [];
        for (let i = 0; i < (props.wordsCount || 1); i++) {
            let lengthWord = Math.floor(Math.random() * 10 + 8);
            words.push(pattern.repeat(lengthWord));
        }
        return words;
    }, [props.wordsCount]);

    return (
        <div aria-busy={true} className={css.container}>
            {text.map((it: string, index: number) => (
                <span
                    key={index}
                    className={cx(['text-placeholder-vars', css.loadingWord, !props.isNotAnimated && css.animatedLoading])}
                    dangerouslySetInnerHTML={{ __html: it }}
                />
            ))}
        </div>
    );
};
