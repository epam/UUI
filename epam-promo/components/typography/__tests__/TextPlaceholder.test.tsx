import React from 'react';
import { TextPlaceholder } from '../TextPlaceholder';
import renderer from 'react-test-renderer';

describe('TextPlaceholder', () => {
    beforeEach(() => {
        jest.spyOn(global.Math, 'random').mockReturnValue(0.5);
    });

    afterEach(() => {
        jest.spyOn(global.Math, 'random').mockRestore();
    });

    it('should be rendered correctly', () => {
        const tree = renderer.create(<TextPlaceholder />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly', () => {
        const tree = renderer.create(<TextPlaceholder color="gray40" wordsCount={5} isNotAnimated />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
