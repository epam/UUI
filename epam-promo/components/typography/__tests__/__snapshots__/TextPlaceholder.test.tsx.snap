// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextPlaceholder should be rendered correctly 1`] = `
<div
  aria-busy={true}
  className="container"
>
  <span
    className="loadingWord text-placeholder-color-gray40 animatedLoading"
    dangerouslySetInnerHTML={
      Object {
        "__html": "0000000000000",
      }
    }
    suppressHydrationWarning={true}
  />
</div>
`;

exports[`TextPlaceholder should be rendered correctly 2`] = `
<div
  aria-busy={true}
  className="container"
>
  <span
    className="loadingWord text-placeholder-color-gray40"
    dangerouslySetInnerHTML={
      Object {
        "__html": "0000000000000",
      }
    }
    suppressHydrationWarning={true}
  />
  <span
    className="loadingWord text-placeholder-color-gray40"
    dangerouslySetInnerHTML={
      Object {
        "__html": "0000000000000",
      }
    }
    suppressHydrationWarning={true}
  />
  <span
    className="loadingWord text-placeholder-color-gray40"
    dangerouslySetInnerHTML={
      Object {
        "__html": "0000000000000",
      }
    }
    suppressHydrationWarning={true}
  />
  <span
    className="loadingWord text-placeholder-color-gray40"
    dangerouslySetInnerHTML={
      Object {
        "__html": "0000000000000",
      }
    }
    suppressHydrationWarning={true}
  />
  <span
    className="loadingWord text-placeholder-color-gray40"
    dangerouslySetInnerHTML={
      Object {
        "__html": "0000000000000",
      }
    }
    suppressHydrationWarning={true}
  />
</div>
`;
